Python/pystate.c

```diff
@@ -815,21 +876,25 @@ new_threadstate(PyInterpreterState *interp)
     if (old_head == NULL) {
         // It's the interpreter's initial thread state.
         assert(id == 1);
+        assert(interp->threads._preallocated_used == 0);
 
-        tstate = alloc_threadstate();
-        if (tstate == NULL) {
-            goto error;
-        }
+        tstate = &interp->_preallocated.tstate;
```

Comment:
This seems like unnecessary complexity. The main interpreter has a pre-allocated tstate, so would hit this code.
The other interpreters will need to dynamically allocate a tstate, so why treat the first thread as special?

Reply:
All interpreters have an initial thread (which will likely be its only thread).

> The other interpreters will need to dynamically allocate a tstate

The first thread state is statically allocated for all interpreters, not just the main interpreter.

Reply:
How is that possible? You don't know how many interpreters there will be.

Reply:
Sorry for the confusion.  I meant that the initial thread state for each `PyInterpreterState` is allocated as part of the interpreter state.  The main interpreter is statically allocated as part of `_PyRuntime` so its initial thread state is likewise statically allocated.  Other interpreters are dynamically allocated, with their initial thread state being a part of that allocation.  We are not pre-allocating thread states separately.